system/lib/fetch/emscripten_fetch.cpp

```diff
@@ -67,16 +67,49 @@ emscripten_fetch_t *emscripten_fetch(emscripten_fetch_attr_t *fetch_attr, const
 	}
 
 	emscripten_fetch_t *fetch = (emscripten_fetch_t *)malloc(sizeof(emscripten_fetch_t));
+	if (!fetch) return 0;
 	memset(fetch, 0, sizeof(emscripten_fetch_t));
 	fetch->id = globalFetchIdCounter++; // TODO: make this thread-safe!
 	fetch->userData = fetch_attr->userData;
-	fetch->url = strdup(url); // TODO: free
+	fetch->url = strdup(url);
 	fetch->__attributes = *fetch_attr;
-	fetch->__attributes.destinationPath = fetch->__attributes.destinationPath ? strdup(fetch->__attributes.destinationPath) : 0; // TODO: free
-	fetch->__attributes.userName = fetch->__attributes.userName ? strdup(fetch->__attributes.userName) : 0; // TODO: free
-	fetch->__attributes.password = fetch->__attributes.password ? strdup(fetch->__attributes.password) : 0; // TODO: free
-	fetch->__attributes.requestHeaders = 0;// TODO:strdup(fetch->__attributes.requestHeaders);
-	fetch->__attributes.overriddenMimeType = fetch->__attributes.overriddenMimeType ? strdup(fetch->__attributes.overriddenMimeType) : 0; // TODO: free
+	fetch->__attributes.destinationPath = fetch->__attributes.destinationPath ? strdup(fetch->__attributes.destinationPath) : 0;
+	fetch->__attributes.userName = fetch->__attributes.userName ? strdup(fetch->__attributes.userName) : 0;
+	fetch->__attributes.password = fetch->__attributes.password ? strdup(fetch->__attributes.password) : 0;
+	if (fetch->__attributes.requestHeaders)
+	{
+		size_t headersCount;
+		for(headersCount = 0; fetch->__attributes.requestHeaders[headersCount]; ++headersCount);
+		const char** headers = (const char**)malloc((headersCount + 1) * sizeof(const char*));
+		if(!headers)
+		{
+			emscripten_fetch_free(fetch);
+			return 0;
+		}
+		for(size_t i = 0; i < headersCount; ++i)
+		{
+			headers[i] = strdup(fetch->__attributes.requestHeaders[i]);
+			if(!headers[i])
+			{
+				emscripten_fetch_free(fetch);
+				return 0;
+			}
+		}
+		headers[headersCount] = 0;
+		fetch->__attributes.requestHeaders = headers;
+	}
+	fetch->__attributes.overriddenMimeType = fetch->__attributes.overriddenMimeType ? strdup(fetch->__attributes.overriddenMimeType) : 0;
+	if (fetch->__attributes.requestData && fetch->__attributes.requestDataSize)
+	{
+		char* data = (char*)malloc(fetch->__attributes.requestDataSize);
```

Comment:
Ouch, this line makes me regret: when POSTing to upload data, the data size can potentially be very large, so it is possible that this `malloc` is on the order of dozens to even up to ~hundred of megabytes in size, and could imaginably fail more easily than the other allocs, since while the fetch is happening, this means we'll be temporarily creating two copies of the uploadable data into the Wasm HEAP.

That makes me think that we should define the semantics of the `requestData` pointer specially that the caller should keep that pointer alive throughout the duration of the call, or otherwise large uploads can run into trouble.

I see that I have documented the `requestData` field as follows in `fetch.h`:

```c
	// If non-zero, specifies a pointer to the data that is to be passed as the body (payload) of the request
	// that is being performed. Leave as zero if no request body needs to be sent.
	// The memory pointed to by this field is provided by the user, and needs to be valid only until the call to
	// emscripten_fetch() returns.
	const char *requestData;
```

however I never had time to implement this field originally, so I think we still have a chance to fix this and make `requestData` be special since it's a potentially fat block of data:

```c
	// If non-zero, specifies a pointer to the data that is to be passed as the body (payload) of the request
	// that is being performed. Leave as zero if no request body needs to be sent.
	// The memory pointed to by this field is provided by the user, and needs to be valid throughout the
        // duration of the fetch operation. If passing a non-zero pointer into this field, make sure to implement
        // the onsuccess and onerror handlers to be notified when the fetch finishes to know when this memory
        // block can be freed. Do not pass a pointer to memory on the stack or other temporary area here.
	const char *requestData;
```

What do you think?